lib/storage/metadata/mongoclient/MongoClientInterface.js

```diff
@@ -1634,6 +1635,103 @@ class MongoClientInterface {
+
+    /*
+     * delete an object that matches a given query object
+     */
+    deleteObjectWithQuery(bucketName, objName, params, log, cb) {
+        const c = this.getCollection(bucketName);
+        const method = 'deleteObjectNoVerConditional';
+        const filter = { _id: objName };
+        try {
+            MongoUtils.translateQuery(0, 'value', filter, params.query);
+        } catch (err) {
+            log.error('error creating mongodb filter', {
+                error: reshapeExceptionError(err),
+            });
+            return cb(errors.InternalError);
+        }
+        return c.findOneAndDelete(filter, (err, res) => {
+            if (err) {
+                log.error('error occurred when attempting to delete object', {
+                    method,
+                    error: err.message,
+                });
+                return cb(errors.InternalError);
+            }
+            if (res.ok !== 1) {
+                log.error('failed to delete object', {
+                    method,
+                    error: err.message,
+                });
+                return cb(errors.InternalError);
+            }
+            /*
+             * unable to find an object that matches the query
+             */
+            if (!res.value) {
+                log.debug('unable to find target object to delete', {
+                    method,
+                    filter,
+                });
+                return cb(errors.NoSuchKey);
+            }
+            return cb();
+        });
+    }
+
+    /*
+     * update an object that matches a given query. If one cannot be found,
+     * a new object will be inserted
+     */
+    putObjectWithQuery(bucketName, objName, objVal, params, log, cb) {
+        const c = this.getCollection(bucketName);
+        const method = 'putObjectNoVerConditional';
+        const filter = { _id: objName };
+        try {
+            MongoUtils.translateQuery(0, 'value', filter, params.query);
+        } catch (err) {
+            log.error('error creating mongodb filter', {
+                error: reshapeExceptionError(err),
+            });
+            return cb(errors.InternalError);
+        }
+        return c.findOneAndUpdate(filter, {
+            $set: {
+                _id: objName,
+                value: objVal,
+            },
+        }, {
+            upsert: true,
+        }, (err, res) => {
+            if (err) {
+                log.error('error occurred when attempting update object', {
+                    method,
+                    error: err,
+                });
+                return cb(errors.InternalError);
+            }
+            if (res.ok !== 1) {
+                log.error('failed to update object', {
+                    method,
+                    error: err.message,
+                });
+                return cb(errors.InternalError);
+            }
+            if (!res.value) {
+                log.debug('object not found...inserting object', {
```

Comment:
Since the action is already done, the log should use past tense to avoid confusion (can be surprising when troubleshooting that the DB does not reflect what the log says).